bliss/simulator/simulated_dataset.py

```diff
@@ -87,3 +91,74 @@ def val_dataloader(self):
 
     def test_dataloader(self):
         return DataLoader(self, batch_size=None, num_workers=self.num_workers)
+
+
+class CachedSimulatedDataset(pl.LightningDataModule, IterableDataset):
+    def __init__(
+        self,
+        n_batches: int,
+        valid_n_batches: int,
+        fix_validation_set: bool,
+        num_workers: int,
+        batch_size: int,
+        file_data_capacity: int,
+        cached_data_path: str,
+    ):
+        super().__init__()
+
+        self.n_batches = n_batches
+        self.valid_n_batches = valid_n_batches
+        self.fix_validation_set = fix_validation_set
+        self.num_workers = num_workers
+        self.batch_size = batch_size
+        self.file_data_capacity = file_data_capacity
+        self.cached_data_path = cached_data_path
+
+        self.data: List[FileDatum] = []
+
+        # assume cached image files exist, read from disk
+        for filename in os.listdir(self.cached_data_path):
+            if "valid" in filename:
+                continue
+            if filename.startswith("dataset") and filename.endswith(".pkl"):
+                self.data += self.read_file(f"{self.cached_data_path}/{filename}")
+        assert self.data, "No cached data loaded; run `generate.py` first"
+
+    def read_file(self, filename: str) -> List[FileDatum]:
+        with open(filename, "rb") as f:
+            return pickle.load(f)
+
+    def get_batch(self, batch_idx) -> Dict:
+        batch_data = self.data[batch_idx * self.batch_size : (batch_idx + 1) * self.batch_size]
+        images = torch.stack([datum["images"] for datum in batch_data])
```

Comment:
If we need this code to be more efficient (i.e., to keep up with the GPU), we should consider storing the images together in memory by putting them all one giant tensor. (Same for the background and the tile catalogs.) As is, we're storing references to the images consecutively, but not the images themselves. So we still end up accessing memory in a fairly random order. Which may or may not be fast enough.

Reply:
I see, yea this makes sense. You mean to say upon shuffling (the references), the accesses are spatially distant, and we are not exploiting spatial locality when we could be?

Reply:
Yes, that's what I had in mind.

Another idea: we could leave it as is an try using multiple workers after all. Not a lot, maybe 2 or 4. Given all the 4 loops in `get_batch` we may benefit from a few workers after all.

Reply:
It would be more elegant to exploit spatial locality, but whatever keeps the GPU busy is fine :-)